### ISSUE TITLE
[SP9[NixOS] Keyboard&Trackpad not working after upgrade to 6.12.4

### ISSUE THREAD

**Issue by dev-nis** [linux-surface/linux-surface#1645]
[Description of the bug or feature]
After upgrading to Surface Kernel 6.12.4 (via nix-hardware-overlay), the Surface Typecover keyboard and mouse don't work anymore. I use disk encryption and upon booting to the decrypting screen, I get two error messages:
```
[        2.070709] surface_serial_hub serial0-0: error -EINVAL: failed to setup IRQ
[        2.071665] surface_serial_hub serial0-0: probe with driver surface_serial_hub failed with error -22
```
All other modules load fine, but with these two fails, the keyboard and mouse don't work. If I unlock the device with an external keyboard and boot up, the typecover keyboard and mouse are still not detected.

If I downgrade to an older kernel version (6.11), it works again.

### Environment

- Hardware model: Surface Pro 9 
- Kernel version: Linux nsclaptop 6.12.4 #1-NixOS SMP PREEMPT_DYNAMIC Mon Dec  9 09:41:16 UTC 2024 x86_64 GNU/Linux
- Distribution: NixOS 24.11

**Comment by dev-nis** [comment]
Here is the `dmesg` : 

[dmesg.txt](https://github.com/user-attachments/files/18277739/dmesg.txt)

**Comment by thmichel** [comment]
Same error here running ublue on a Surface Laptop Studio. Keyboard and touchpad don't work.

**Comment by sargon** [comment]
Same here.

I have bisected a little bit more and also tested the latest? version of the 6.11 kernel branch supported in linux-surface/kernel (still building, edit when finished).

N 6.12.4 -> 6.12.7
? 6.11.11 
Y 6.11.4

The main difference (on this repo) between 6.11 and 6.12 patchsets is the removal of [0007-surface-sam.patch](https://github.com/linux-surface/linux-surface/blob/master/patches/6.11/0007-surface-sam.patch)  and renumbering of the following patches. My understanding is that this patch is now part of the upstream kernel and thus no longer needed
(Have not checked/compared the versions here (6.11) and in upstream) .

There are changes between 6.11/0010-surface-typecover.patch and 6.12/0009-surface-typecover.patch maybe this is a good point for further investigation.

### Environment

* Hardware model: Surface Pro 8

**Comment by dev-nis** [comment]
6.11.11 works for me

**Comment by sargon** [comment]
> There are also changes between 6.11/0010-surface-typecover.patch and 6.12/0009-surface-typecover.patch, maybe that is a good point for further investigation.

I inspected the changes and their is only change in the surrounding patched code. And to my understanding we are not affected by those.

My conclusion so far is that we are affected by some changes in upstream kernel.

**Comment by sargon** [comment]
6.11.11 works for me, but display-manager is not starting properly.
6. 12.1 does not work

**Comment by sargon** [comment]
6.12-rc1 does not work

**Comment by sargon** [comment]
I read a little bit kernel code and git blamed the code base a little I suspect 
[b27622f](https://github.com/torvalds/linux/commit/b27622f1317271b88048ff2d76fead28b72aeccf#diff-80690b7b074f126d25d3f56018857a6b157558ba74770ec4ca4ae7da17c5f59f) to be the breaking change. I try to build a kernel without this patch tomorrow.

**Comment by qzed** [comment]
@sargon I doubt it's that commit, there should not be any functional changes regarding x86-based devices. Could you post your kernel config?

**Comment by qzed** [comment]
Based on the dmesg log above, the pinctrl drivers are loaded later than the SAM module:
```
[    3.389556] surface_serial_hub serial0-0: SAM firmware version: 6.89.139
[    3.390584] surface_serial_hub serial0-0: error -EINVAL: failed to setup IRQ
[    3.390925] surface_serial_hub serial0-0: probe with driver surface_serial_hub failed with error -22
[    3.392593] stage-1-init: [Mon Dec 30 21:00:56 UTC 2024] loading module pinctrl_tigerlake...
[    3.416275] stage-1-init: [Mon Dec 30 21:00:56 UTC 2024] loading module btrfs...
```
Please make sure that you are building in the pinctrl modules. I've been told that that is generally expected on x86 systems, and not having them built in can also break other things (like volume buttons).

**Comment by sargon** [comment]
Thank you very much. That did the trick.

The module was build, but wasn't in the modules list and so not loaded infront of the other modules.
After I put it there works just fine.

**Comment by sargon** [comment]
One thing keeps wondering me. Why not putting the module into the list wasn't a problem in linux < 6.12

**Comment by dev-nis** [comment]
> Thank you very much. That did the trick.
> 
> The module was build, but wasn't in the modules list upfront cryptsetup and so not loaded infront of the other modules. After I put it there works just fine.

What did you exactly do? Because i'd like to also solve this problem

**Comment by sargon** [comment]
> What did you exactly do? Because i'd like to also solve this problem

I moved the module `pinctrl_tigerlake` up in the list, but I don't really understand why it worked previously, when it was at the bottom of the list.

  initrd.kernelModules = [
      "pinctrl_tigerlake"
      "surface_aggregator"
      "surface_aggregator_registry"
      "surface_aggregator_hub"
      "surface_hid_core"
      "8250_dw"
      "surface_hid"
      "intel_lpss"
      "intel_lpss_pci"
    ];

**Comment by sargon** [comment]
> One thing keeps me wondering. Why not putting the module into the list wasn't a problem in linux < 6.12

I checked the modules.dep of the old kernel version and compared it to the new kernel version - no difference.
And in both versions i don't see a direct or indirect module dependencie between surface modules and the pinctrl modules.
I checked the initrd generation code of nixos the parts of the module loading is >3y old so there is no change, too.
My guess is that some timing (some module waits on somethings) has changed and thus the parallel loading has changed.... 

@qzed I think [Wiki: Disk Encryption](https://github.com/linux-surface/linux-surface/wiki/Disk-Encryption) should mention that GPIO modules should be loaded first.

**Comment by dev-nis** [comment]
I can confirm that just reordering the modules in `initrd.kernelModules` worked! Thanks for your help! 
I think this issue can be closed. Otherwise, feel free to reopen it :)

**Comment by qzed** [comment]
> I checked the modules.dep of the old kernel version and compared it to the new kernel version - no difference. And in both versions i don't see a direct or indirect module dependencie between surface modules and the pinctrl modules.

Yeah, there is no explicit dependency because the systems are somewhat parallel. So the SAM module depends on the pinctrl/GPIO interface. But it doesn't (and shouldn't) know what's going on beyond that interface, as in which driver is actually responsible for controlling GPIO pins. So it can't depend on the specific driver that's needed to enable that interface.

There are ways to defer probing the driver if GPIO pins aren't available, but I've been told that I should rather expect them to be available and that any GPIO pin drivers should be directly built into the kernel. Maybe we need to re-evaluate that. 

> I checked the initrd generation code of nixos the parts of the module loading is >3y old so there is no change, too. My guess is that some timing (some module waits on somethings) has changed and thus the parallel loading has changed....

Yes, could be something in the timing the modules are loaded changed. Ideally you would build in the respective GPIO driver into the kernel image itself, so it doesn't have to be loaded as module

> Never mind, edited myself

Thanks!
